src/main/java/cubicchunks/worldgen/generator/vanilla/VanillaCompatibilityGenerator.java

```diff
@@ -0,0 +1,216 @@
+import cubicchunks.worldgen.generator.ICubeGenerator;
+import cubicchunks.worldgen.generator.ICubePrimer;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.entity.EnumCreatureType;
+import net.minecraft.init.Blocks;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3i;
+import net.minecraft.world.World;
+import net.minecraft.world.WorldProviderHell;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.biome.Biome.SpawnListEntry;
+import net.minecraft.world.chunk.Chunk;
+import net.minecraft.world.chunk.IChunkGenerator;
+import net.minecraft.world.chunk.storage.ExtendedBlockStorage;
+import net.minecraftforge.fml.common.registry.GameRegistry;
+
+public class VanillaCompatibilityGenerator implements ICubeGenerator {
+
+	private IChunkGenerator vanilla;
+	private ICubicWorld world;
+
+	private Chunk lastChunk;
+	private boolean optimizationHack;
+
+	private boolean stripBadrock;
+	private IBlockState badrock = Blocks.BEDROCK.getDefaultState();
+	private IBlockState underBlock = Blocks.STONE.getDefaultState();
+
+	public VanillaCompatibilityGenerator(IChunkGenerator vanilla, ICubicWorld world) {
+		this.vanilla = vanilla;
+		this.world = world;
+
+		// heuristics TODO: add a config that overrides this
+		lastChunk = vanilla.provideChunk(0, 0); // lets scan the chunk at 0, 0
+
+		IBlockState topstate = null;
+		int         topcount = 0;
+		{   // find the type of block that is most common on the bottom layer
+			IBlockState laststate = null;
+			for(int at = 0;at < 16 * 16;at++){
+				IBlockState state = lastChunk.getBlockState(at | 0x0F, 0, at >> 4);
+				if(state != laststate){
+
+					int count = 1;
+					for(int i = at + 1;i < 16 * 16;i++){
+						if(lastChunk.getBlockState(i | 0x0F, 0, i >> 4) == state){
+							count++;
+						}
+					}
+					if(count > topcount){
+						topcount = count;
+						topstate = state;
+					}
+				}
+				laststate = state;
+			}
+		}
+
+		if(topstate.getBlock() != Blocks.BEDROCK){
+			underBlock = topstate;
+		}else{
+			stripBadrock = true;
+			underBlock = world.getProvider() instanceof WorldProviderHell
+					? Blocks.NETHERRACK.getDefaultState()
+					: Blocks.STONE.getDefaultState(); //TODO: maybe scan for stone type?
+		}
+	}
+
+	private Biome[] biomes;
+	@Override
+	public void generateColumn(Column column) {
+
+		this.biomes = this.world.getBiomeProvider()
+				.getBiomes(this.biomes, 
+						Coords.cubeToMinBlock(column.getX()),
+						Coords.cubeToMinBlock(column.getZ()),
+						Coords.CUBE_MAX_X, Coords.CUBE_MAX_Z);
+
+		byte[] abyte = column.getBiomeArray();
+        for (int i = 0; i < abyte.length; ++i)
+        {
+            abyte[i] = (byte)Biome.getIdForBiome(this.biomes[i]);
+        }
+	}
+
+	@Override
+	public void recreateStructures(Column column) {
+		vanilla.recreateStructures(column, column.getX(), column.getZ());
+	}
+
+	@Override
+	public ICubePrimer generateCube(int cubeX, int cubeY, int cubeZ) {
```

Comment:
And generation is one of the places where it would make sense to use CubeCoords. And with CubeCoords you can add a method like `cubePos.forEachBlock((x, y, z)->code;)`